keps/sig-api-machinery/3617-fine-grained-authz/README.md

```diff
@@ -0,0 +1,1162 @@
+or with in-tree verbs. We may not know exactly what the best practices are until
+we have an alpha or beta behind us, hence we're not ready to specify that in
+this KEP yet.
+
+(The exact permission system will be described below because we're not done with
+the requirements yet, keep reading!)
+
+#### Uncovered verbs
+
+There's one more piece of information we need to satisfy our second goal, and it
+can't go in the schema because it may vary from cluster to cluster.
+
+This concept is whether a permission verb is "covered" by enclosing permissions,
+or whether it is "uncovered". If a users tries to make a change, and one of the
+fields changing has a permission verb which is "uncovered", the user must have
+that permission in order to make the change, it's no longer good enough to have
+the general update permission or even an enclosing permission (like permission
+to modify metadata).
+
+To support this, we will add a new resource to the system. Each object will
+contain a list of permissions to mark as "uncovered". We support multiple
+objects to make it easy for us to add some default object(s) in the future, and
+to let cluster admins mark certain permissions as uncovered in their particular
+cluster.
+
+The k8s project can probably only permit brand new fields to have "uncovered"
+permissions by default, otherwise we would break existing users. But cluster
+admins know the exact authz system they use, and they can simultanesouly mark a
+permission as "uncovered" while granting it via their authz system, so that
+there is no outage.
+
+Example:
+
+```
+apiVersion: permissionconfig.k8s.io/v1
+kind: UncoveredPermissionSet
+metadata:
+  name: legacySensitiveLabels
+spec:
+  uncoveredPermissions:
+  - "granular:label(pod-security.kubernetes.io)" # for the enforce label
+  - "granular:label(scheduler.alpha.kubernetes.io)" # for the node-selector label
+```
+
+#### API Server Additional Plumbing and Checks
+
+Currently apiserver performs the authz check before descending into any detailed
+logic for mutating operations. This means that when it is doing the authz check,
+it knows the *object* that is being changed, but nothing about *which fields*
+are changing.
+
+We will introduce a new permission, "granular". This permission
+means that although the caller doesn't have generic update permissions, it may
+still change some fields if it has the specific permissions required.
+
+API Server already maintains schema information for all resource types. This
+will be watched to also maintain field <-> permission mappings.
+
+API Server will be modified to watch UncoveredPermissionSets, to keep a map in
+memory of these.
+
+Now we can describe the modifications to API Server:
+
+On any mutating call,
+
+1. Check the "PUT" / "PATCH" / "CREATE" permission. If the actor has this,
+   place a marker=regular in the request context and proceed to step 3.
+2. Otherwise, check the "granular" permission. If yes, place a
```

Comment:
Does a `granular` check make sense in the context of a create? It seems reasonable to me that it would only apply to update/patch requests.

Reply:
You might let people make a Pod, provided they don't set `schedulingGates`, and then have exceptions - maybe one principal can set _this particular `schedulingGate`_ but not an arbitrary one.

Similar idea to setting a finalizer at create time. If we want to have fine grained control, I think it could be frustrating to have to duplicate the logic / whatever so that we restrict both creates and updates.

Reply:
Yeah the intention is that the per-field permissions are the same for both, otherwise I have to duplicate this all to handle both paths.

The result is you won't be able to create things if you lack a granular permission on a required, non-defaulted field.

Reply:
Does this check apply before or after static defaults?

Reply:
I think after? And probably some extra work not described in the KEP needs to be done to make it work like I just said.

I hope to make another pass on this KEP soon

Reply:
So if a field needs an uncovered permission, it *cannot* have a default value (or it has to be defaulted in registry, not static) - is that the right implication?